### DISCUSSION TITLE
Queries regarding discussion forum APIs for integration with Sunbird coKreat BB

### DISCUSSION THREAD

**Discussion by joffinjoy** [Sunbird-Lern/Community#27] [Query]
With regard to: https://github.com/orgs/Sunbird-inQuiry/discussions/8#discussioncomment-3160667

We want to leverage discussion forum APIs for a review comment functionality (for questionSets) in Sunbird coKreat BB. As such we have few queries regarding the same. 

1. As mentioned in the above linked comment, Hierarchy in discussion forum is suggested to be **category -> forum -> topic -> posts**, but in the **Sunbird Lern Discussion Forum** Features/Core Capabilities page (https://lern.sunbird.org/learn/product-and-developer-guide/discussion-forum/features-core-capabilities) the exact hierarchy isn't shown, as in, the **forum** is missing from the image posted. So we are confused about the exact hierarchy and how each layer interact with each other. The page hints at the idea of having multiple categories and multiple discussions under forums. So wouldn't the forum be at the top of the hierarchy?

2. Next one is regarding the users in discussion forums. The above linked comment do mention users and privileges. But can a user create a post without having to create an account in nodeBB? Meaning, can we pass the simple userId along with the post creation APIs and they are stored without any nodeBB-specific context. 

   What we have found is that in-order for a user to be created in nodeBB, we have to first use the creation API which takes in the username/email and password etc (Which leverages **nodeBB write-API**). But in our use-case, we can't do this since the discussion APIs are abstracted away from the user. The user shouldn't have to signup anywhere (_That too for a second time_) just to just add reviewer and creator comments. How do we tackle this requirement? 

   And at what layer do we add users? Category or Forums or Topics? And can user make a post without being added to any of the above layers first?

3. Third question is regarding the custom tags. Can we fetch posts/categories/topics etc by the user created custom tags that are attached with them? This wasn't clear from the existing discussion APIs mentioned at https://lern.sunbird.org/learn/product-and-developer-guide/discussion-forum/api-documentation . And at what all layers can we attach tags?

4. The **Sunbird Lern Discussion Forum** Features/Core Capabilities page mentions replies to a post. And https://lern.sunbird.org/learn/product-and-developer-guide/discussion-forum/api-documentation/post-apis shows that both **post** creation and **reply to post** creation have the same endpoints and HTTP method. So are the replies just posts with mandatory toPid (Id of the parent post I presume)? And can this extend to any number of levels? (Post -> Reply -> Reply To Reply -> Reply... etc)

cc: @aks30 @vijiurs, @kiranharidas187 @aishwaryashikshalokam, @priyanka-TL, @VISHNUDAS-tunerlabs, @Shakthieshwari, @Prateek-slokam, @MeghnaSL, @githubpiyush @pradeepkumarcm96

**Comment by VenkatKadiveti** [comment]
Hi @joffinjoy 

1. The exact hierarchy of the discussion forum is  Category => Sub Categories (Optional) => Topics => Posts. What we have in the document image ([link](https://lern.sunbird.org/learn/product-and-developer-guide/discussion-forum/features-core-capabilities)).
2. For Nodebb user creation - Without any user account in nodebb, we can't perform any write operation.
	1. Topic creation.
	2. Reply to a topic (post).
	3. Reply to reply (post).
	4. upvote/downvote.
	5. bookmark ... etc
Suppose you want to perform any of these operations. You must pass _uid (in a body or as a query param in URL)  along with the Master token as the Authorization token in headers. if not you will get a 401 error.

3. We can get the topics/posts based on the tag name. By using the below API 
`
curl --location --request POST '{{host}}/discussions/api/tags/list' \
--header 'Authorization: Bearer {{Nodebb Master Token}}' \
--header 'Content-Type: application/json' \
--header 'Cookie: express.sid=s%3Az9kvyUoCkBurqP9xmxg9I_FbYB7bUm2u.z%2Fio7jM%2FEVxNq3YnyaeQcsWWqeP43Af7qNLg3l8BXGY' \
--data-raw '{
    "request": {
        "cid": ["1663"], // Category id 
        "tag": "tag-name" // Tag name
    }
}'
`
 
4. We can create replies at any level.
`
Topic => reply to topic (Post1) =>  reply to Post1 (Post2) => reply to Post2 (Post3)........ so on
`
CC: @reshmi-nair

**Reply by joffinjoy** [reply]
> For Nodebb user creation - Without any user account in nodebb, we can't perform any write operation.

So currently how do you integrate nodeBB with other building blocks? I mean, if we have to create user accounts in nodebb, wouldn't the users have to do the signup twice? Once for the BB and once again for nodeBB?

Currently, we already have user accounts (in Sunbird coKreat), and want to leverage discussion forum APIs for a review flow. We wouldn't have an actual discussion forum UI (Like the one shown in the video here: https://lern.sunbird.org/learn/product-and-developer-guide/discussion-forum). 

How can we avoid multiple signup for the same user? 

For context: https://github.com/orgs/Sunbird-inQuiry/discussions/8#discussioncomment-3195228

**Reply by VenkatKadiveti** [reply]
@joffinjoy 

Currently, we are creating a new account in nodebb. and this will happen when you click on the **Fourn Icon** we are making the user create call to create a user in nodebb if the user already exists we are just rupturing the user details.

Answer to your question 

> So currently how do you integrate nodeBB with other building blocks? I mean, if we have to create user accounts in nodebb, wouldn't the users have to do the signup twice? Once for the BB and once again for nodeBB?

Yes, We are creating 2 accounts one is for BB and another is for nodebb.

CC: @reshmi-nair

**Comment by VenkatKadiveti** [comment]
Hello @joffinjoy 

1. If you see [here](https://github.com/orgs/Sunbird-Lern/discussions/27#discussioncomment-3196775) I have mentioned sub-category is Optional, which Means If we want we can create or else not.  But the main flow is Category => Topic => Post.
2. A forum is nothing but a category that we created in the nodebb and attached to the context. Here the main thing is whatever the category that we created we have to link it with the context. 
 Examples:
	- Group context => {type: 'group', identifier: 'group-id'}
	- Course Contest => {type: 'course', identifier: 'course-id'}
	- Batch context => {type: 'bathc', identifier: 'batch-id'}
	- questionset context => {type: 'questionset', identifier: 'questionset-id'}
 
We can use the API `discussion/forum/v3/create` to create a forum [Attaching category for context](https://lern.sunbird.org/learn/product-and-developer-guide/discussion-forum/api-documentation/forum-apis). 

3. When you launched into the discussion forum, we are calling the below APIs

- 	 [Get Forum Id](https://lern.sunbird.org/learn/product-and-developer-guide/discussion-forum/api-documentation/forum-apis) by passing context and it will return cid along with context data
`Get Forum ID Response data
---------------------------
{
  "id": ".discussions.api.forum.v2.read",
  "ver": "1.0",
  "ets": 1658559298361,
  "params": {
        "resmsgid": "267822f1-087b-476e-bc31-c55245b35df2",
        "msgid": "",
        "status": "Success"
     },
  "responseCode": "OK",
  "result": [
          {
              "sbIdentifier": "d1b25f1a-7ba4-404e-87bb-56f605ad4a86",
              "sbType": "group",
              "cid": "1774"
          }
       ]
}`

- Once we get the response from the above API, we are taking the cid and calling  [Get the category details by category id](https://lern.sunbird.org/learn/product-and-developer-guide/discussion-forum/api-documentation/category-apis)  this will give you all the topics which are created under the category.

3.  I have shared the curl `discussion/tags/list` to fetch topics and posts based on tag name under the selected category.

CC: @reshmi-nair @rhwarrier
